### PR TITLE
Add useMemo, tests, and prettier (unformatted)

### PR DESCRIPTION
**Pull request by tannerlinsley**

### CODE REVIEW

**Comment by tannerlinsley** [comment]
Took some git selection foo, but this should do the trick. If, and when, you're ready you can run `yarn format` to adopt prettier styling to the repo.

**Comment by getify** [comment]
Thanks!

**Comment by tannerlinsley** [comment]
I think it's worth noting that the `useCallback` hook can now also be implemented via proxy of the `useMemo` hook:

```javascript
function useCallback (fn, ...guards) {
	return useMemo(() => fn, ...guards)
}
```

File: package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
 		"test:dist": "TEST_DIST=true npm test",
 		"test:package": "TEST_PACKAGE=true npm test",
 		"test:all": "npm test && npm run test:dist && npm run test:package",
+		"format": "prettier README.md {src,scripts,tests}/**/*.js --write",
 		"coverage": "istanbul cover scripts/node-tests.js",
```
Comment (getify) [inline package.json]: Let's leave prettier out of TNG for the purposes of this PR.

File: package.json [hunk 2]
```diff
@@ -18,6 +19,7 @@
 	},
 	"devDependencies": {
 		"coveralls": "~3.0.2",
+		"prettier": "^1.15.3",
```
Comment (getify) [inline package.json]: Let's leave prettier out of TNG for the purposes of this PR.

File: src/tng-hooks.src.js [hunk 3]
```diff
@@ -8,8 +8,8 @@
 	var buckets = new WeakMap();
 	var tngStack = [];
 
-	return { TNG, useState, useReducer, useEffect, };
 
+	return { TNG, useState, useReducer, useEffect, useMemo };
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
	return { TNG, useState, useReducer, useEffect, useMemo, };
```

Comment (getify) [reply src/tng-hooks.src.js]: I prefer trailing commas in most lists.

File: src/tng-hooks.src.js [hunk 4]
```diff
@@ -213,4 +224,45 @@
...
+			if (guardsChanged(memoization[1], guards)) {
+				// invoke the memoization
+				memoization[0] = fn();
+				memoization[1] = guards;
+			}
+
+			bucket.nextMemoizationIdx++;
+
+			return memoization[0];
+		} else {
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
		}
		else {
```

File: tests/tests.js [hunk 5]
```diff
@@ -1,12 +1,13 @@
 "use strict";
 
 QUnit.test( "API", function test(assert){
-	assert.expect( 4 );
+	assert.expect(5);
```
Comment (getify) [inline tests/tests.js]:
```suggestion
	assert.expect( 5 );
```

File: tests/tests.js [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
 
 	assert.ok( _isFunction( TNG ), "TNG()" );
 	assert.ok( _isFunction( useState ), "useState()" );
 	assert.ok( _isFunction( useReducer ), "useReducer()" );
 	assert.ok( _isFunction( useEffect ), "useEffect()" );
+	assert.ok(_isFunction(useMemo), "useMemo()");
```
Comment (getify) [inline tests/tests.js]:
```suggestion
	assert.ok( _isFunction(useMemo), "useMemo()" );
```

File: tests/tests.js [hunk 7]
```diff
@@ -236,6 +237,43 @@ QUnit.test( "useEffect(..)", function test(assert){
 	assert.strictEqual( pActual, pExpected, "call without TNG wrapping context" );
 } );
 
+QUnit.test("useMemo(..)", function test(assert) {
+	function foo(num, rest) {
+		return useMemo(() => {
+			return num;
```
Comment (getify) [inline tests/tests.js]: These tests don't really *prove* the caching is happening. I'd prefer to use `assert.step(..)` (see other tests) to verify which times the memoized function was invoked vs not. That way we're really sure the memoization is happening.

File: tests/tests.js [hunk 8]
```diff
@@ -236,6 +237,43 @@ QUnit.test( "useEffect(..)", function test(assert){
 	assert.strictEqual( pActual, pExpected, "call without TNG wrapping context" );
 } );
 
+QUnit.test("useMemo(..)", function test(assert) {
+	function foo(num, rest) {
+		return useMemo(() => {
```
Comment (getify) [inline tests/tests.js]:
```suggestion
		return useMemo( function memoized(){
```

Comment (getify) [reply tests/tests.js]: No arrow functions in this test suite.

File: tests/tests.js [hunk 9]
```diff
@@ -236,6 +237,43 @@ QUnit.test( "useEffect(..)", function test(assert){
...
+			return num;
+		}, rest);
+	}
+
+	function baz(num, rest) {
+		return useMemo(() => {
```
Comment (getify) [inline tests/tests.js]:
```suggestion
		return useMemo( function memoized(){
```

File: tests/tests.js [hunk 10]
```diff
@@ -236,6 +237,43 @@ QUnit.test( "useEffect(..)", function test(assert){
...
+			return num * num;
+		}, rest);
+	}
+
+	var pExpected = "error";
```
Comment (getify) [inline tests/tests.js]: I realize this test suite style is rather unorthodox. It's habit from a long time ago for me. See the other tests, where I use the expected/actual pattern, with like `rExpected` and `rActual`. For reference, I use the peculiar progression of naming prefixes for these pairs of variables of `r`, `p`, `q`, `t`, `s`, `u`, `h`, and `j`.

I am happy to fix up this test accordingly, but if you look at the other tests you should be able to see the pattern if you want to give it a shot.

File: tests/tests.js [hunk 11]
```diff
@@ -236,6 +237,43 @@ QUnit.test( "useEffect(..)", function test(assert){
...
+	assert.strictEqual(foo(8, []), 8, "useMemo = 8");
+
+	try {
+		var pActual = baz();
+	} catch (err) {
+		var pActual = "error";
+	}
+
+	assert.strictEqual(pActual, pExpected, "call without TNG wrapping context");
+	assert.expect(9);
```
Comment (getify) [inline tests/tests.js]: `assert.expect(..)` should come at the top of the assertions block.

File: tests/tests.js [hunk 12]
```diff
@@ -236,6 +237,43 @@ QUnit.test( "useEffect(..)", function test(assert){
...
+});
```
Comment (getify) [inline tests/tests.js]:
```suggestion
} );
```

File: src/tng-hooks.src.js [hunk 13]
```diff
@@ -213,4 +224,45 @@
...
+			throw new Error(
```
Comment (getify) [inline src/tng-hooks.src.js]: let's move this back to a single line statement for now.

File: src/tng-hooks.src.js [hunk 14]
```diff
@@ -84,8 +87,16 @@
...
-				bucket = { nextStateSlotIdx: 0, nextEffectIdx: 0, stateSlots: [], effects: [], cleanups: [], };
-				buckets.set(tngf,bucket);
+				bucket = {
+					nextStateSlotIdx: 0,
+					nextEffectIdx: 0,
+					nextMemoizationIdx: 0,
+					stateSlots: [],
+					effects: [],
+					cleanups: [],
+					memoizations: []
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
					memoizations: [],
```

File: src/tng-hooks.src.js [hunk 15]
```diff
@@ -84,8 +87,16 @@
...
+				};
+				buckets.set(tngf, bucket);
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
				buckets.set(tngf,bucket);
```

File: src/tng-hooks.src.js [hunk 16]
```diff
@@ -213,4 +224,45 @@
 			throw new Error("useEffect() only valid inside an Articulated Function or a Custom Hook.");
 		}
 	}
+
+	function useMemo(fn, ...guards) {
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
	function useMemo(fn,...guards) {
```

File: src/tng-hooks.src.js [hunk 17]
```diff
@@ -213,4 +224,45 @@
...
+		}
+
+		var bucket = getCurrentBucket();
+		if (bucket) {
+			// need to create this memoization-slot for this bucket?
+			if (!(bucket.nextMemoizationIdx in bucket.memoizations)) {
+				bucket.memoizations[bucket.nextMemoizationIdx] = [];
+			}
+
+			let memoizationIdx = bucket.nextMemoizationIdx;
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
```

Comment (getify) [reply src/tng-hooks.src.js]: We're not using an internal closure (like we are with `useEffect(..)`) so no need to create this variable for this hook.

File: src/tng-hooks.src.js [hunk 18]
```diff
@@ -213,4 +224,45 @@
...
+			let memoization = bucket.memoizations[memoizationIdx];
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
			let memoization = bucket.memoizations[bucket.nextMemoizationIdx];
```

File: src/tng-hooks.src.js [hunk 19]
```diff
@@ -213,4 +224,45 @@
...
+
+			// check guards?
+			if (guardsChanged(memoization[1], guards)) {
+				// invoke the memoization
+				memoization[0] = fn();
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
 	 	 	 	try {
				 	memoization[0] = fn();
 	 	 	 	}
 	 	 	  	finally {
```

File: src/tng-hooks.src.js [hunk 20]
```diff
@@ -213,4 +224,45 @@
...
+				memoization[1] = guards;
```
Comment (getify) [inline src/tng-hooks.src.js]:
```suggestion
				 	memoization[1] = guards;
				 }
```

File: src/tng-hooks.src.js (hunk 19 above)
Comment (getify) [reply src/tng-hooks.src.js]: Note that this is the first place where some of the changes I mentioned in the main review message above would go. Line 256 here would likely look like `memoization[0] = fn;` (not `fn();`).

File: src/tng-hooks.src.js [hunk 21]
```diff
@@ -213,4 +224,45 @@
...
+			}
+
+			bucket.nextMemoizationIdx++;
+
+			return memoization[0];
```
Comment (getify) [inline src/tng-hooks.src.js]: Note that this is where most of the changes from the main review message would go. This line 261 should make a return a function that checks if the value in the slot of `memoization[0]` had been called yet or not, invoke if it not, replace that slot with its return value, and then subsequently just keep returning that value from then on.

**Comment by tannerlinsley** [comment]
Can you point me to the react hooks source you're referring to?

**Comment by getify** [comment]
Oh, oops. Sorry, I think I got confused with [the `useCallback(..)` hook](https://reactjs.org/docs/hooks-reference.html#usecallback), which does indeed return a new wrapped function, not the result of calling the provided function. IIUC, your suggested implementation above for `useCallback(..)` wouldn't be correct, because your version calls `useMemo(..)` right away, it doesn't return a wrapped function.

----

However, the docs here do suggest a bit of behavior for `useMemo(..)` that I don't think is currently handled by your implementation:

https://reactjs.org/docs/hooks-reference.html#usememo

> If no array is provided, a new value will be computed whenever a new function instance is passed as the first argument. (With an inline function, on every render.)

IOW, it seems like if no guards list is passed, you have to put a reference to the original function *as* the only value in the guards list, and make the guards comparison accordingly each time.

**Comment by tannerlinsley** [comment]
Yeah I can see that in the react hooks source now. Simple enough. My original comment on how to implement useCallback is still valid. Both hooks use the same mechanism to determine when a new value is returned. The only difference is that useCallback returns the function you pass as the first argument, instead of executing and returning its result. Clearly you could duplicate the useMemo code and make that single change, but I say why waste?

**Comment by getify** [comment]
> useCallback returns the function you pass as the first argument

I don't see how that could possibly be right. What `useCallback(..)` is supposed to return, according to the docs, is a memoized function, not just the original function.

Can you link me to this implementation code?

**Comment by tannerlinsley** [comment]
https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberHooks.js#L609

**Comment by getify** [comment]
Thanks, I got it now. Sorry for my confusion on that stuff.

**Comment by tannerlinsley** [comment]
No worries. Hooks are still very fresh, so it's easy to lose track of what does what and how. Most of my understanding came from writing my use-react-hooks packport for older react versions.

**Comment by getify** [comment]
Just wanted to check, would you like to make the suggested updates here, or would you like me to pull this into a fork and do them before merging?

**Comment by getify** [comment]
@tannerlinsley Can I ask one more thing of you? Can you squash all the commits into a single one (I guess maybe with a new PR), so that when I pull that into a branch to fix up, it has a single commit attributed to you? Otherwise, I think it's going to erase your attribution.

**Comment by tannerlinsley** [comment]
If you go into your repo settings and enable the `sqaush` merge option for PR's, then all you need to do is select that option on this PR and click merge. All attribution will be retained.

**Comment by getify** [comment]
OK, cool, that can work.

I was going to pull this into a separate branch first, do any extra fixes to tests and some extra documentation there, and then merge back to master. The squash merge will go directly to master, though. I wonder if there's a way to redirect to a separate branch?

**Comment by getify** [comment]
Merged `useMemo(..)` into master just now (after cleaning up tests and docs). will release it as 4.1.0, but after I add the `useRef(..)` hook.

Thanks again!

**Comment by tannerlinsley** [comment]
You bet Kyle. Good luck!

**Pull request merged**